README.md
Translatable Routing
=================
[![Build Status](https://travis-ci.org/webbuilders-group/silverstripe-translatablerouting.png)](https://travis-ci.org/webbuilders-group/silverstripe-translatablerouting)  ![helpfulrobot](https://helpfulrobot.io/webbuilders-group/silverstripe-translatablerouting/badge)

Extends SilverStripe Translatable module and replaces routing to enable multi-lingual URLs. By default the URLs that will be generated with this module installed will look like ``http://example.com/en/`` though there are settings for using the locale instead of just the language code (see [the documentation](docs/en/configuration.md) for information).

## Maintainer Contact
* Ed Chipman ([UndefinedOffset](https://github.com/UndefinedOffset))

## Requirements
* SilverStripe CMS 3.2.x+
* [SilverStripe Translatable](https://github.com/silverstripe/silverstripe-translatable/) 2.1+


## Installation
__Composer (recommended):__
```
composer require webbuilders-group/silverstripe-translatablerouting
```

If you prefer you may also install manually:
* Download the module from here https://github.com/webbuilders-group/silverstripe-translatablerouting/archive/master.zip
* Extract the downloaded archive into your site root so that the destination folder is called translatablerouting, opening the extracted folder should contain _config.php in the root along with other files/folders
* Run dev/build?flush=all to regenerate the manifest

After installing you must make some modifications to your Page class, [see here](docs/en/usage.md) for more information.


## Documentation
For full usage and configuration documentation see the [docs folder](docs/en/).


## Notes
Translatable Routing has support for the SilverStripe Google Sitemaps module for 3.1, which will add support for the multi-lingual site per [google's documentation](https://support.google.com/webmasters/answer/2620865?hl=en) on doing this.


CHANGELOG.md
# Change Log

## [0.5.0](https://github.com/webbuilders-group/silverstripe-translatablerouting/tree/0.5.0) (2017-10-10)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-translatablerouting/compare/0.4.1...0.5.0)

**Implemented enhancements:**

- Expand MultilingualModelAsControllerTest to cover more routing [\#21](https://github.com/webbuilders-group/silverstripe-translatablerouting/issues/21)
- Improve documentation to meet SilverStripe Module Standards [\#18](https://github.com/webbuilders-group/silverstripe-translatablerouting/issues/18)
- Support for country only urls [\#17](https://github.com/webbuilders-group/silverstripe-translatablerouting/issues/17)
- Improve module description in readme to include sample urls [\#9](https://github.com/webbuilders-group/silverstripe-translatablerouting/issues/9)

**Fixed bugs:**

- Build Failure in MultilingualModelAsControllerTest::testMultilingualRequired [\#22](https://github.com/webbuilders-group/silverstripe-translatablerouting/issues/22)
- MultilingualControllerTest does not cover dash locales [\#20](https://github.com/webbuilders-group/silverstripe-translatablerouting/issues/20)
- Extensions do not appear to take into account the MultilingualRootURLController.UseDashLocale setting [\#19](https://github.com/webbuilders-group/silverstripe-translatablerouting/issues/19)
- Locale url redirect incorrect when using dash locales [\#11](https://github.com/webbuilders-group/silverstripe-translatablerouting/issues/11)

## [0.4.1](https://github.com/webbuilders-group/silverstripe-translatablerouting/tree/0.4.1) (2016-03-23)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-translatablerouting/compare/0.4.0...0.4.1)

**Merged pull requests:**

- Added standard Scrutinizer config [\#16](https://github.com/webbuilders-group/silverstripe-translatablerouting/pull/16) ([helpfulrobot](https://github.com/helpfulrobot))
- Added standard .gitattributes file [\#15](https://github.com/webbuilders-group/silverstripe-translatablerouting/pull/15) ([helpfulrobot](https://github.com/helpfulrobot))
- Added standard .editorconfig file [\#13](https://github.com/webbuilders-group/silverstripe-translatablerouting/pull/13) ([helpfulrobot](https://github.com/helpfulrobot))
- FIX: composer package name [\#12](https://github.com/webbuilders-group/silverstripe-translatablerouting/pull/12) ([christopherdarling](https://github.com/christopherdarling))

## [0.4.0](https://github.com/webbuilders-group/silverstripe-translatablerouting/tree/0.4.0) (2014-10-19)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-translatablerouting/compare/0.3.2...0.4.0)

**Implemented enhancements:**

- Add note about Translatable.enforce\_global\_unique\_urls [\#7](https://github.com/webbuilders-group/silverstripe-translatablerouting/issues/7)
- Support for locale's with a dash instead of an underscore [\#4](https://github.com/webbuilders-group/silverstripe-translatablerouting/issues/4)

**Fixed bugs:**

- HTTP\_ACCEPT\_LANGUAGE may not be present [\#8](https://github.com/webbuilders-group/silverstripe-translatablerouting/issues/8)
- Test using deprecated methods [\#6](https://github.com/webbuilders-group/silverstripe-translatablerouting/issues/6)

## [0.3.2](https://github.com/webbuilders-group/silverstripe-translatablerouting/tree/0.3.2) (2014-07-17)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-translatablerouting/compare/0.1.7...0.3.2)

## [0.1.7](https://github.com/webbuilders-group/silverstripe-translatablerouting/tree/0.1.7) (2014-07-17)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-translatablerouting/compare/0.1.6...0.1.7)

**Fixed bugs:**

- Google sitemap hreflang broken [\#5](https://github.com/webbuilders-group/silverstripe-translatablerouting/issues/5)

## [0.1.6](https://github.com/webbuilders-group/silverstripe-translatablerouting/tree/0.1.6) (2014-06-10)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-translatablerouting/compare/0.3.1...0.1.6)

## [0.3.1](https://github.com/webbuilders-group/silverstripe-translatablerouting/tree/0.3.1) (2014-06-10)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-translatablerouting/compare/0.1.5...0.3.1)

**Fixed bugs:**

- Redirect loop when parent page is unpublished [\#2](https://github.com/webbuilders-group/silverstripe-translatablerouting/issues/2)

## [0.1.5](https://github.com/webbuilders-group/silverstripe-translatablerouting/tree/0.1.5) (2014-03-11)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-translatablerouting/compare/0.3.0...0.1.5)

## [0.3.0](https://github.com/webbuilders-group/silverstripe-translatablerouting/tree/0.3.0) (2014-03-11)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-translatablerouting/compare/0.1.4...0.3.0)

**Closed issues:**

- Incorrect language on controllers [\#1](https://github.com/webbuilders-group/silverstripe-translatablerouting/issues/1)

## [0.1.4](https://github.com/webbuilders-group/silverstripe-translatablerouting/tree/0.1.4) (2014-02-13)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-translatablerouting/compare/0.2.1...0.1.4)

## [0.2.1](https://github.com/webbuilders-group/silverstripe-translatablerouting/tree/0.2.1) (2014-02-13)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-translatablerouting/compare/0.2.0...0.2.1)

## [0.2.0](https://github.com/webbuilders-group/silverstripe-translatablerouting/tree/0.2.0) (2014-01-10)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-translatablerouting/compare/0.1.3...0.2.0)

## [0.1.3](https://github.com/webbuilders-group/silverstripe-translatablerouting/tree/0.1.3) (2013-09-09)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-translatablerouting/compare/0.1.2...0.1.3)


tests/pre-up/travis_setup.php
<?php
if (php_sapi_name() != 'cli') {
    header('HTTP/1.0 404 Not Found');
    exit;
}

$opts = getopt('', array(
	'target:', // required
));

// Sanity checks
if(!$opts || !isset($opts['target'])) {
	echo "Invalid arguments specified\n";
	exit(1);
}

$targetPath=$opts['target'];
if(!file_exists("$targetPath/mysite/code/Page.php")) {
    echo "Cannot find Page class in mysite\n";
    exit(1);
}

$pageDOContent='
    public function Link($action=null) {
        if(MultilingualRootURLController::config()->use_country_only) {
            $i18nSegment=strtolower(preg_replace(\'/^(.*?)_(.*?)$/\', \'$2\', $this->Locale));
        }else if(MultilingualRootURLController::config()->UseLocaleURL) {
            if(MultilingualRootURLController::config()->UseDashLocale) {
                $i18nSegment=str_replace(\'_\', \'-\', strtolower($this->Locale));
            }else {
                $i18nSegment=$this->Locale;
            }
        }else {
            $i18nSegment=i18n::get_lang_from_locale($this->Locale);
        }
        
        return Controller::join_links(Director::baseURL(), $i18nSegment.\'/\', $this->RelativeLink($action));
    }
    
    public function RelativeLink($action=null) {
        if($this->ParentID && self::nested_urls()) {
            $base=$this->Parent()->RelativeLink($this->URLSegment);
        } else {
            $base=$this->URLSegment;
        }

        //Unset base for homepage URLSegments in their default language. Homepages with action parameters or in different languages need to retain their URLSegment. We can only do this if the homepage is on the root level.
        if(!$action && $base==Translatable::get_homepage_link_by_locale($this->Locale) && !$this->ParentID) {
            $base = null;
        }

        return Controller::join_links($base, \'/\', $action);
    }
}';

$pageControllerContent='
        if($this->dataRecord && $this->dataRecord instanceof SiteTree && MultilingualRootURLController::should_be_on_root($this->dataRecord) && (!isset($this->urlParams[\'Action\']) || !$this->urlParams[\'Action\']) && !$_POST && !$_FILES && !$this->redirectedTo()) {
            $getVars=$_GET;
            unset($getVars[\'url\']);

            if(MultilingualRootURLController::config()->use_country_only) {
                $i18nSegment=strtolower(preg_replace(\'/^(.*?)_(.*?)$/\', \'$2\', $this->Locale));
            }else if(MultilingualRootURLController::config()->UseLocaleURL) {
                if(MultilingualRootURLController::config()->UseDashLocale) {
                    $i18nSegment=str_replace(\'_\', \'-\', strtolower($this->Locale));
                }else {
                    $i18nSegment=$this->Locale;
                }
            }else {
                $i18nSegment=i18n::get_lang_from_locale($this->Locale);
            }
            
            
            if($getVars) {
                $url=$i18nSegment.\'/?\'.http_build_query($getVars);
            }else {
                $url=$i18nSegment.\'/\';
            }

            $this->redirect($url, 301);
            return;
        }';

        
//Update Page.php
$pageContents=file_get_contents("$targetPath/mysite/code/Page.php");
$pageContents=preg_replace('/\}/s', $pageDOContent, $pageContents);

$f=fopen("$targetPath/mysite/code/Page.php", 'w');
fwrite($f, $pageContents);
fclose($f);


//Update page controller content
$pageContents=file_get_contents("$targetPath/mysite/code/Page_Controller.php");
$pageContents=preg_replace('/public function init\(\)(\s+)\{(.*?)\}(.*?)\}/s', 'public function init()$1{$2 '.$pageControllerContent."\n}\n}", $pageContents);

$f=fopen("$targetPath/mysite/code/Page_Controller.php", 'w');
fwrite($f, $pageContents);
fclose($f);
?>

tests/MultilingualControllerTest.php
<?php
class MultilingualControllerTest extends FunctionalTest {
    private $origLocale;
    private $origCurrentLocale;
    private $origAllowedLocales;
    private $origi18nLocale;
    private $origCookieLocale;
    private $origAcceptLanguage;
    private $origLocaleRoutingEnabled;
    private $origDashLocaleEnabled;
    private $origCountryOnly;
    
    protected $autoFollowRedirection=false;
    
    public function setUp() {
        parent::setUp();
        
        $this->origCountryOnly=MultilingualRootURLController::config()->use_country_only;
        MultilingualRootURLController::config()->use_country_only=false;
        
        $this->origLocaleRoutingEnabled=MultilingualRootURLController::config()->UseLocaleURL;
        MultilingualRootURLController::config()->UseLocaleURL=false;
        
        $this->origDashLocaleEnabled=MultilingualRootURLController::config()->UseDashLocale;
        MultilingualRootURLController::config()->UseDashLocale=false;
        
        $this->origAcceptLanguage=$_SERVER['HTTP_ACCEPT_LANGUAGE'];
        $_SERVER['HTTP_ACCEPT_LANGUAGE']='en-US,en;q=0.5';
        
        $this->origCookieLocale=Cookie::get('language');
        Cookie::force_expiry('language');
        Cookie::set('language', '');
        
        $this->origCurrentLocale=Translatable::get_current_locale();
        Translatable::set_current_locale('en_US');
        
        $this->origLocale=Translatable::default_locale();
        Translatable::set_default_locale('en_US');
        
        $this->origi18nLocale=i18n::get_locale();
        i18n::set_locale('en_US');
        
        $this->origAllowedLocales=Translatable::get_allowed_locales();
        Translatable::set_allowed_locales(array('en_US', 'fr_FR'));
        
        
        MultilingualTestController::add_extension('MultilingualControllerExtension');
        
        Config::inst()->update('Director', 'rules', array(
                                                        '$Language/multilingual-test-controller//$Action/$ID/$OtherID'=>'MultilingualTestController'
                                                    ));
        
        MultilingualRootURLController::reset();
    }
    
    public function tearDown() {
        MultilingualRootURLController::config()->use_country_only=$this->origCountryOnly;
        MultilingualRootURLController::config()->UseLocaleURL=$this->origLocaleRoutingEnabled;
        MultilingualRootURLController::config()->UseDashLocale=$this->origDashLocaleEnabled;
        
        Translatable::set_current_locale($this->origCurrentLocale);
        Translatable::set_default_locale($this->origLocale);
        Translatable::set_allowed_locales($this->origAllowedLocales);
        
        i18n::set_locale($this->origi18nLocale);
        
        Cookie::force_expiry('language');
        
        if($this->origCookieLocale) {
            Cookie::set('language', $this->origCookieLocale);
        }
        
        $_SERVER['HTTP_ACCEPT_LANGUAGE']=$this->origAcceptLanguage;
        
        MultilingualRootURLController::reset();
        
        parent::tearDown();
    }
    
    /**
     * Tests to see if the controller responds correctly if the language is in the url
     */
    public function testLanguageInURL() {
        //Perform Request
        $response=$this->get('fr/multilingual-test-controller/');
        
        
        //Ensure a 200 response
        $this->assertEquals(200, $response->getStatusCode());
        
        
        //Verify the response matches what is expected
        $this->assertEquals('i18n: fr_FR|Translatable: fr_FR', $response->getBody());
    }
    
    /**
     * Tests to see if the controller responds correctly if the locale is in the url
     */
    public function testLocaleInURL() {
        //Enable locale urls
        MultilingualRootURLController::config()->UseLocaleURL=true;
    
        //Set accept language to french
        $_SERVER['HTTP_ACCEPT_LANGUAGE']='fr-FR,fr;q=0.5';
    
        //Get the root url
        $response=$this->get('fr_FR/multilingual-test-controller/');
        
        
        //Ensure a 200 response
        $this->assertEquals(200, $response->getStatusCode());
        
        
        //Verify the response matches what is expected
        $this->assertEquals('i18n: fr_FR|Translatable: fr_FR', $response->getBody());
    }
    
    /**
     * Tests to see if the controller responds correctly if the dashed locale is in the url
     */
    public function testDashLocaleInURL() {
        //Enable locale urls
        MultilingualRootURLController::config()->UseLocaleURL=true;
        MultilingualRootURLController::config()->UseDashLocale=true;
    
        //Set accept language to french
        $_SERVER['HTTP_ACCEPT_LANGUAGE']='fr-FR,fr;q=0.5';
    
        //Get the root url
        $response=$this->get('fr-fr/multilingual-test-controller/');
        
        
        //Ensure a 200 response
        $this->assertEquals(200, $response->getStatusCode());
        
        
        //Verify the response matches what is expected
        $this->assertEquals('i18n: fr_FR|Translatable: fr_FR', $response->getBody());
    }
    
    /**
     * Tests to see if the controller responds correctly if the county is in the url
     */
    public function testCountryInURL() {
        //Enable locale urls
        MultilingualRootURLController::config()->use_country_only=true;
        
        //Set accept language to french
        $_SERVER['HTTP_ACCEPT_LANGUAGE']='fr-FR,fr;q=0.5';
        
        
        //Change the allowed locales to have fr_CA
        Translatable::set_allowed_locales(array('en_US', 'fr_CA'));
        
        
        //Get the root url
        $response=$this->get('ca/multilingual-test-controller/');
        
        
        //Ensure a 200 response
        $this->assertEquals(200, $response->getStatusCode());
        
        
        //Verify the response matches what is expected
        $this->assertEquals('i18n: fr_CA|Translatable: fr_CA', $response->getBody());
    }
    
    /**
     * Tests to see if the controller responds correctly if the language is in the url
     */
    public function testAutoDetectLanguage() {
        //Set accept language to french
        $_SERVER['HTTP_ACCEPT_LANGUAGE']='fr-FR,fr;q=0.5';
        Translatable::set_default_locale('fr_FR');
        Translatable::set_current_locale('fr_FR');
        i18n::set_locale('fr_FR');
        
        
        //Perform Request
        $response=$this->get('MultilingualTestController');
        
        
        //Ensure a 200 response
        $this->assertEquals(200, $response->getStatusCode());
        
        
        //Verify the response matches what is expected
        $this->assertEquals('i18n: fr_FR|Translatable: fr_FR', $response->getBody());
    }
    
    /**
     * Tests to see if the controller responds correctly if the language is in the url
     */
    public function testInvalidLanguageURL() {
        //Perform Request
        $response=$this->get('es/multilingual-test-controller/');
        
        
        //Ensure a 301 response
        $this->assertEquals(301, $response->getStatusCode());
        
        
        //Verify the redirect url matches the default
        $this->assertEquals(Director::baseURL().'en/multilingual-test-controller', $response->getHeader('Location'));
    }
}

class MultilingualTestController extends Controller implements TestOnly {
    private static $extensions=array(
                                    'MultilingualControllerExtension'
                                );
    
    public function index() {
        return 'i18n: '.i18n::get_locale().'|Translatable: '.Translatable::get_current_locale();
    }
}
?>

code/control/MultilingualGoogleSitemap.php
<?php
class MultilingualGoogleSitemapController extends GoogleSitemapController
{
    private static $allowed_actions=array(
                                            'sitemap'
                                        );
    
    /**
     * Specific controller action for displaying a particular list of links
     * for a class
     *
     * @return mixed
     */
    public function sitemap()
    {
        if ($this->request->param('ID')=='SiteTree') {
            //Disable the locale filter
            Translatable::disable_locale_filter();
            
            
            $items=parent::sitemap();
            
            
            //Re-enable the locale filter
            Translatable::enable_locale_filter();
            
            
            return $items;
        } else {
            return parent::sitemap();
        }
    }
}


code/extensions/MultilingualSiteTreeExtension.php
<?php
class MultilingualSiteTreeExtension extends DataExtension {
    /**
     * Updates the CMS fields adding the fields defined in this extension
     * @param {FieldList} $fields Field List that new fields will be added to
     */
    public function updateCMSFields(FieldList $fields) {
        $urlSegmentField=$fields->dataFieldByName('URLSegment');
        if($urlSegmentField) {
            if(MultilingualRootURLController::config()->use_country_only) {
                $i18nSegment=strtolower(preg_replace('/^(.*?)_(.*?)$/', '$2', $this->owner->Locale));
            }else if(MultilingualRootURLController::config()->UseLocaleURL) {
                if(MultilingualRootURLController::config()->UseDashLocale) {
                    $i18nSegment=str_replace('_', '-', strtolower($this->owner->Locale));
                }else {
                    $i18nSegment=$this->owner->Locale;
                }
            }else {
                $i18nSegment=i18n::get_lang_from_locale($this->owner->Locale);
            }
            
            
            $baseLink=Controller::join_links(
                                            Director::absoluteBaseURL(),
                                            $i18nSegment.'/',
                                            (SiteTree::config()->nested_urls && $this->owner->ParentID ? $this->owner->Parent()->RelativeLink(true):null)
                                        );
            
            
            $urlSegmentField->setURLPrefix($baseLink);
        }
    }
    
    /**
     * Gets the RFC 1766 version of the input locale
     * @return {string} RFC 1766 version of the locale (i.e en-US)
     */
    public function getRFC1766Locale() {
        return i18n::convert_rfc1766($this->owner->Locale);
    }
}
?>

code/control/MultilingualRootURLController.php
<?php
class MultilingualRootURLController extends RootURLController {
    /**
     * Tells the routing controllers to use the locale url or not i.e /en_US/ instead of /en/
     * @default false
     * @config MultilingualRootURLController.UseLocaleURL
     */
    private static $UseLocaleURL=false;
    
    /**
     * Tells the routing controllers to use a dashed locale or not i.e /en-us/ instead of /en_US/ requires MultilingualRootURLController.UseLocaleURL to be true
     * @default false
     * @config MultilingualRootURLController.UseDashLocale
     */
    private static $UseDashLocale=false;
    
    /**
     * 
     * @default false
     * @config MultilingualRootURLController.use_country_only
     */
    private static $use_country_only=false;
    
    
    public function handleRequest(SS_HTTPRequest $request, DataModel $model=null) {
        self::$is_at_root=true;
        $this->setDataModel($model);
        
        $this->pushCurrent();
        $this->init();
        
        //Get the locale from the language param
        if($language=$request->param('Language')) {
            if(MultilingualRootURLController::config()->use_country_only) {
                $locale=self::get_locale_from_country($request->param('Language'));
                if(empty($locale)) {
                    //Locale not found 404
                    try {
                        $response=ErrorPage::response_for(404);
                        $this->httpError(404, ($response ? $response:'The requested page could not be found.'));
                    }catch(SS_HTTPResponse_Exception $responseException) {
                        $result=$responseException->getResponse();
                    }
                    
                    return $result;
                }
            }else if(MultilingualRootURLController::config()->UseLocaleURL) {
                if(MultilingualRootURLController::config()->UseDashLocale) {
                    //Language is missing a dash 404
                    if(strpos($language, '-')===false) {
                        //Locale not found 404
                        try {
                            $response=ErrorPage::response_for(404);
                            $this->httpError(404, ($response ? $response:'The requested page could not be found.'));
                        }catch(SS_HTTPResponse_Exception $responseException) {
                            $result=$responseException->getResponse();
                        }
                        
                        return $result;
                    }
                
                    $locale=explode('-', $language);
                    $locale[1]=strtoupper($locale[1]);
                    
                    //Make sure that the language is all lowercase
                    if($language==implode('-', $locale)) {
                        //Locale not found 404
                        try {
                            $response=ErrorPage::response_for(404);
                            $this->httpError(404, ($response ? $response:'The requested page could not be found.'));
                        }catch(SS_HTTPResponse_Exception $responseException) {
                            $result=$responseException->getResponse();
                        }
                        
                        return $result;
                    }
                    
                    $locale=implode('_', $locale);
                }else {
                    //Language is missing an underscore 404
                    if(strpos($language, '_')===false) {
                        //Locale not found 404
                        try {
                            $response=ErrorPage::response_for(404);
                            $this->httpError(404, ($response ? $response:'The requested page could not be found.'));
                        }catch(SS_HTTPResponse_Exception $responseException) {
                            $result=$responseException->getResponse();
                        }
                        
                        return $result;
                    }
                    
                    $locale=$language;
                }
            }else if(strpos($language, '_')!==false || strpos($language, '-')!==false) {//If the url has a locale in it when the settings are off
                //Locale not found 404
                try {
                    $response=ErrorPage::response_for(404);
                    $this->httpError(404, ($response ? $response:'The requested page could not be found.'));
                }catch(SS_HTTPResponse_Exception $responseException) {
                    $result=$responseException->getResponse();
                }
                
                return $result;
            }else {//Potentially a language code
                $locale=i18n::get_locale_from_lang($language);
            }
            
            if(in_array($locale, Translatable::get_allowed_locales())) {
                Cookie::set('language', $locale);
                
                Translatable::set_current_locale($locale);
                i18n::set_locale($locale);
                
                
                if(!DB::isActive() || !ClassInfo::hasTable('SiteTree')) {
                    $this->response=new SS_HTTPResponse();
                    $this->response->redirect(Director::absoluteBaseURL().'dev/build?returnURL='.(isset($_GET['url']) ? urlencode($_GET['url']):null));
                    return $this->response;
                }
                
                $request->setUrl($language.'/'.self::get_homepage_link().'/');
                $request->match('$Language/$URLSegment//$Action', true);
                
                
                $controller=new MultilingualModelAsController();
                $result=$controller->handleRequest($request, $model);
                
                $this->popCurrent();
                return $result;
            }else {
                $locale=Translatable::default_locale();
                if(MultilingualRootURLController::config()->UseLocaleURL || MultilingualRootURLController::config()->use_country_only) {
                    if(MultilingualRootURLController::config()->UseDashLocale) {
                        $language=str_replace('_', '-', strtolower($locale));
                    }else {
                        $language=$locale;
                    }
                    
                    //For country only the language code in the url should be just the country code
                    if(MultilingualRootURLController::config()->use_country_only) {
                        $language=strtolower(preg_replace('/^(.*?)_(.*?)$/', '$2', $language));
                    }
                }else {
                    $language=Translatable::default_lang();
                }
                
                
                //URL Param Locale is not allowed so redirect to default
                $this->redirect(Controller::join_links(Director::baseURL(), $language).'/', 301);
                
                $this->popCurrent();
                return $this->response;
            }
        }
        
        
        //No Locale Param so detect browser language and redirect
        if($locale=self::detect_browser_locale()) {
            if(MultilingualRootURLController::config()->use_country_only) {
                $language=$locale;
            }else if(MultilingualRootURLController::config()->UseLocaleURL || MultilingualRootURLController::config()->use_country_only) {
                if(MultilingualRootURLController::config()->UseDashLocale) {
                    $language=str_replace('_', '-', strtolower($locale));
                }else {
                    $language=$locale;
                }
            }else {
                $language=i18n::get_lang_from_locale($locale);
            }
            
            
            Cookie::set('language', $locale);
            
            
            //For country only the language code in the url should be just the country code
            if(MultilingualRootURLController::config()->use_country_only) {
                $language=strtolower(preg_replace('/^(.*?)_(.*?)$/', '$2', $language));
            }
            
            
            $this->redirect(Controller::join_links(Director::baseURL(), $language).'/', 301);
            
            $this->popCurrent();
            return $this->response;
        }
        
        
        if(MultilingualRootURLController::config()->UseLocaleURL || MultilingualRootURLController::config()->use_country_only) {
            if(MultilingualRootURLController::config()->UseDashLocale) {
                $language=str_replace('_', '-', strtolower(Translatable::default_locale()));
            }else {
                $language=Translatable::default_locale();
            }
            
            //For country only the language code in the url should be just the country code
            if(MultilingualRootURLController::config()->use_country_only) {
                $language=strtolower(preg_replace('/^(.*?)_(.*?)$/', '$2', $language));
            }
        }else {
            $language=Translatable::default_lang();
        }
        
        $this->redirect(Controller::join_links(Director::baseURL(), $language.'/'), 301);
        
        $this->popCurrent();
        return $this->response;
    }
    
    /**
     * Determines the locale best matching the given list of browser locales
     * @return {string} The matching locale, or null if none could be determined
     */
    public static function detect_browser_locale() {
        if($locale=Cookie::get('language')) {
            if(in_array($locale, Translatable::get_allowed_locales())) {
                return $locale;
            }else {
                Cookie::force_expiry('language');
            }
        }
        
        // Given multiple canditates, narrow down the final result using the client's preferred languages
        $inputLocales=(array_key_exists('HTTP_ACCEPT_LANGUAGE', $_SERVER) ? $_SERVER['HTTP_ACCEPT_LANGUAGE']:null);
        if(empty($inputLocales)) {
            return null;
        }
        
        // Generate mapping of priority => list of languages at this priority
        // break up string into pieces (languages and q factors)
        preg_match_all('/(?<code>[a-z]{1,8}(-[a-z]{1,8})?)\s*(;\s*q\s*=\s*(?<priority>1|0\.[0-9]+))?/i', $inputLocales, $parsedLocales);
        
        $prioritisedLocales=array();
        if(count($parsedLocales['code'])) {
            // create a list like "en" => 0.8
            $parsedLocales=array_combine($parsedLocales['code'], $parsedLocales['priority']);
    
            // Generate nested list of priorities => [languages]
            foreach($parsedLocales as $language => $priority) {
                $priority=(empty($priority) ? 1.0:floatval($priority));
                if(empty($prioritisedLocales[$priority])) {
                    $prioritisedLocales[$priority] = array();
                }
                
                $prioritisedLocales[$priority][]=$language;
            }
                
            // sort list based on value
            krsort($prioritisedLocales, SORT_NUMERIC);
        }
    
        // Check each requested language against loaded languages
        foreach($prioritisedLocales as $priority=>$parsedLocales) {
            foreach($parsedLocales as $browserLocale) {
                foreach(Translatable::get_allowed_locales() as $language) {
                    if(stripos(preg_replace('/_/', '-', $language), $browserLocale)===0) {
                        return $language;
                    }
                }
            }
        }
    
        return null;
    }
    
    
    
    /**
     * Returns TRUE if a request to a certain page should be redirected to the site root (i.e. if the page acts as the home page).
     * @param {SiteTree} $page
     * @return {bool}
     */
    public static function should_be_on_root(SiteTree $page) {
        if(!self::$is_at_root && self::get_homepage_link()==trim($page->RelativeLink(true), '/')) {
            return true;
        }
        
        return false;
    }
    
    /**
     * Finds the locale based on the allowed locales and the country code
     * @param {string} $country Country to try and find a locale for
     * @return {string} Can potentially return null if there were no matches
     */
    public static function get_locale_from_country($country) {
        $potentialLocales=preg_grep('/_'.preg_quote(strtoupper($country), '/').'$/', Translatable::get_allowed_locales());
        
        
        //Verify there is only one match
        if(count($potentialLocales)>1) {
            user_error('Warning there are multiple allowed locales ("'.implode('", "', $potentialLocales).'") matching the given country "'.strtoupper($country).'" you should not use the MultilingualRootURLController.use_country_only with the current set of allowed locales: "'.implode('", "', Translatable::get_allowed_locales()).'"', E_USER_WARNING);
        }
        
        
        //Pop off the first element in the array
        return array_shift($potentialLocales);
    }
}
?>